Make the output of info_auto be the list of tactics it finds

jfehrle opened the pull request:
- [ ] Added / updated **test-suite**.
- [ ] Added **changelog**.
- [ ] Added / updated **documentation**.

ppedrot commented:
Public service announcement: at some point I'll start drawing a hammer every time somebody uses the word "proof tree" and use it to nail into their brain the fact that **THERE IS NO SUCH THING AS A TACTIC PROOF TREE**. Trying to get a script that "behaves the same" as whatever auto did is a basic no-go, so please don't try to print bulleted proof-y blobs of text to add confusion for the people who still believe that proof trees are a thing.

herbelin commented:
@ppedrot, you're completely right from a theoretical point of view, but isn't there a possible pragmatical compromise? I feel that when we are able to generate an imperative proof script that does the same as `auto` (backtracking steps removed), it is still worth to offer the possibility to show it, and when not, we fail.

herbelin commented:
More generally, `auto` is convenient but it is not fully clear to me that when we interactively use `auto` in a proof script, we really want to keep `auto` rather than whatever emulation of it in the proof script. I wonder what others think.

ppedrot commented:
The auto tactic provides a lot of robustness against future change, even in proof scripts. I don't see the point of wanting to expand it into a tactic script. If you go that way you could even argue that all tactics should be macro-expanded in the source into the proof term they generate.

herbelin commented:
> The auto tactic provides a lot of robustness against future change, even in proof scripts. I don't see the point of wanting to expand it into a tactic script. If you go that way you could even argue that all tactics should be macro-expanded in the source into the proof term they generate.

Yes, it is exactly my question. I rarely use `auto` myself, so I'm missing experience, but when I worked on the synthesis of well-designed hint bases, I observed that the least change was easily breaking `auto`, so I concluded that `auto` was not very robust (with this specific regard). But maybe I'm wrong and `auto` is better than a list of `apply`, I don't know.

jfehrle commented:
@ppedrot Threats of violence, even in jest, have no place in the Coq community.

> every time somebody uses the word "proof tree" and use it to nail into their brain the fact that THERE IS NO SUCH THING AS A TACTIC PROOF TREE.

Why would you expect any user to think otherwise unless it is explained to them?  After all, applying tactics in proofs looks like a tree, and one would expect that auto is essentially doing the same thing.  If you or Hugo provide a working counterexample I'll put this in the documentation, perhaps on the description of `info_auto`.

Even so, the tactics appear to form a tree roughly 100% of the time.  If there are detectable exceptions, the `auto_info` output could indicate the exception, hopefully specifically.  Failing that, the documentation can note that `info_auto` output isn't always accurate.  Nonetheless, users may want to see what auto has done and learn from it, so it's quite useful.  The version output by this PR is much easier to read than `info_auto`.  Do you hate `info_auto` and `info_debug`?  How about `info_eauto`, which gives output that resembles a simple list of tactics?  You're not forced to use them.

> @herbelin More generally, `auto` is convenient but it is not fully clear to me that when we interactively use `auto` in a proof script, we really want to keep `auto` rather than whatever emulation of it in the proof script. I wonder what others think.

It should be the user's choice.  It would be great for a GUI to have function key that invokes proof search (e.g. `auto`) and appends the resulting tactics to the proof script.  Easier to use, also why repeat the proof search, which in some cases may take many seconds or minutes, every time you rerun the script?  Users who don't want this behavior can type `auto`.

> @ppedrot The auto tactic provides a lot of robustness against future change, even in proof scripts.

If a proof later fails, you can easily do more proof search to fix it.  I suppose the proof search-generated tactics could be flagged with comments on the first and last lines.

jfehrle commented:
> @herbelin To get more bullet levels, you can use --, ++, **, ---, +++, etc, or even -+, -+, --+, ...

Great, I'll do that, probably only ones like `---` but not like `--+`.

> knowing from where the hint comes (such as `external` and `in plus`) seems useful to me. However it could indeed be just an extra option to `info_auto`.

Yeah, sometimes users will want to see this info and sometimes not.  For now, I think I'll print both of these together on the right hand side as a comment.  Providing a way to control this can come later.  I think you'd also want to see the id number, provided we fix `Print HintDb` to show it instead of always showing `id = 0`.

I'm curious why we decided to name them `External` hints and why it's worth printing `(* external *)` on every line.  I don't understand why there are so many types of hints--e.g., why are Resolve, Rewrite distinct from External hints?  Also, the documentation on the hint types and the search process is very thin.  We should have a bunch of examples and tips on how to structure hintDbs.

jfehrle commented:
Now the output looks like this:

![image](https://github.com/user-attachments/assets/2cca021c-0eb9-4291-aedb-65c8f79d600b)

With a more complex script, it looks very busy.  Not my favorite.  WDYT?

![image](https://github.com/user-attachments/assets/d100e1b5-477a-437a-805d-fc86cd46e3db)

Zimmi48 commented:
> @ppedrot Threats of violence, even in jest, have no place in the Coq community.

Jim is right. @ppedrot: we are accustomed to your style, so we tend to not be bothered by it, but consider that there are people looking at our public discussions that do not know us and that are from a wide range of diverse cultures. You can't keep expressing yourself like this. This is purely and simply not in line with our code of conduct (cf. in particular the sections "Be careful in the words that you choose" and "Remember that what you write in a public online forum might be read by many people you don't know. Consider what image your words will give to outsiders of the development team / the user community as a whole.").

herbelin commented:
> If you or Hugo provide a working counterexample I'll put this in the documentation, perhaps on the description of `info_auto`.

I would say that there are two different questions: whether a proof is a tree, and when `info_auto` is not correct.

Two typical examples I see where a  proof is not a tree is when you prove a goal by side effect as in:
```coq
Goal exists x, x=0.
unshelve esplit.
2:auto.
Qed.
```
or when, what you have is a forest, as in:
```coq
Goal forall x, x = x /\ x <= x.
intros; split; info_auto.
```
(@ppedrot may have other examples in mind though).

As for an example where bullet-presented `info_auto` does not help, consider e.g.:
```coq
Goal 
Goal forall x, x = x /\ 0 <= x.
intros; split; info_auto; apply le_0_n.
Qed.
```

herbelin commented:
> 'm curious why we decided to name them `External` hints

They can be arbitrary, on the contrary of Resolve.

> and why it's worth printing `(* external *)` on every line.

It is not mandatory,  but it is sometimes useful to know from were come the hint.

> I don't understand why there are so many types of hints--e.g., why is Resolve ~Rewrite~ distinct from External hints? What's the point of Immediate?

Immediate is like Resolve but with the constraint that it does not generate subgoals (iirc).

> Also, the documentation on the hint types and the search process is very thin. We should have a bunch of examples and tips on how to structure hintDbs.

Probably...

jfehrle commented:
Thanks for the examples, but it looks like `info_auto` works correctly for all of them (even without the changes in this PR).  Maybe could be tweaked slightly for the multigoal case.  (Apologies for any mistakes, I have two fingers in a splint.)

**First**
```coq
Goal exists x, x=0.
unshelve esplit.
2:info_auto.

    (* info_auto: *)
    simple apply @eq_refl.  (* in core *)

Qed.
```
Substituting works fine:
```coq
Goal exists x, x=0.
unshelve esplit.
2: simple apply @eq_refl.
Qed.
```

**Second**

(* or when, what you have is a forest, as in: *)
```coq
Goal forall x, x = x /\ x <= x.
intros; split; info_auto.

    (* info auto: *)
    simple apply @eq_refl.  (* in core *)
    (* info auto: *)
    simple apply le_n.  (* in core *)

Qed.
```

Substituting:

```coq
Goal forall x, x = x /\ x <= x.
intros; split.
simple apply @eq_refl.
simple apply le_n.
Qed.
```

**Third**

(* (@ppedrot may have other examples in mind though). *)

(* As for an example where bullet-presented info_auto does not help, consider e.g.: *)

```coq
Goal forall x, x = x /\ 0 <= x.
intros; split; info_auto; apply le_0_n.

    (* info_auto: *)
    simple apply @eq_refl.  (* in core *)
    (* info_auto: *)
    idtac.  
Qed.
```

Substituting:

```coq
Goal forall x, x = x /\ 0 <= x.
intros; split.
simple apply @eq_refl.
idtac.
apply le_0_n. 
Qed.
```

jfehrle commented:
@ppedrot Can we make some progress on this, or at least have a civil discussion?  Please respond.  Thanks.

@herbelin any other ideas on an example of what PM is getting at?

herbelin commented:
> @herbelin any other ideas on an example of what PM is getting at?

I'm a bit out of ideas. @ppedrot would know better what examples to give.

jfehrle commented:
@ppedrot I added the same formatting to `info_eauto` so you can try your example.

ppedrot commented:
Here is a somewhat simple example:
```coq
Create HintDb foo.

Hint Resolve conj : foo.
Hint Resolve eq_refl : foo.

Goal forall n, n=1 -> exists x y : nat, x = y /\ x = 0.
intros.
do 2 eexists; subst.
info_auto with nocore foo.
(*
(* info auto: *)
simple apply conj.  (* in foo *)
- simple apply @eq_refl.  (* in foo *)
- simple apply @eq_refl.  (* in foo *)
*)
Undo.
simple apply conj.
- Fail simple apply @eq_refl.  (* Unable to unify "?y@{n:=1}" with "?x@{n:=1}" ... *)
- simple apply @eq_refl.
```
This failure is due to the fact that conversion constraints are solved by default after every dot but not across auto calls. I reused bits of the examples from #3141.

jfehrle commented:
Interesting example, thanks.  Several thoughts:

If we think there are a bunch of quirks like this, then the doc can include a general disclaimer for `info_auto` output.  If this is the only quirk, then maybe we can explain in a little more detail.

I notice that this works if I solve the second goal first.  It that a good workaround that will always work?

```
simple apply conj.
 2: simple apply @eq_refl.
 simple apply @eq_refl.
```

But IINM this behavior isn't due to a problem with the "proof tree" reconstruction--the output of the `auto_info` old and new is equivalent.

> This failure is due to the fact that conversion constraints are solved by default after every dot but not across auto calls.

Is the difference in behavior correct (e.g. for performance reasons) or maybe a bug?

Also it's interesting that `do 2 eexists; subst` gives different results from `eexists; subst. eexists; subst`.  I'd expect them to behave identically.  Is this just another manifestation of when conversion constraints are solved?

Also, BTW, the doc doesn't even explain the meaning of the last three productions below.  If someone can explain them a little I will add that info.  Does the third one just mean that the substitutions need to be applied once a value is chosen for `?x/?y`?

![image](https://github.com/user-attachments/assets/01cae48b-59ec-4a4b-af1b-3e240c192748)
